Proposal: Database Read Component using QueryBuilder

hendesi opened the pull request:
### Workflow & Motivation

This PR aims to add a query builder which is independent of the used exporter and the query parameters that come with the request. The main goal is to create a `Read` Component which is not constrained to a single query, but is able to execute any query which is defined by the parameters of the request. The functionality of this component depends on the work of [dynamics](https://github.com/Apodini/Apodini/pull/56) by @theMomax. As `Dynamics` is not merged yet and there is some functionality of the query builder needs to be further defined, I would consider this PR a WIP and can used as a place for further discussions. This approach is similar to the one I worked on with @nerdsupremacist and presented on the intermediate presentation, but as for now, I think I could simply it a bit.

The `QueryBuilder` has a `func info(for type: Model.Type) -> [ModelInfo]` which returns an array of `ModelInfo`, where every element represents one property of the database model. This property contains field key and its type. The `Read` Component uses this info to initialise the `@Dynamics`:

```
public init() {
        var dynamicValues: [String: Param] = [:]
        let infos = QueryBuilder.info(for: Model.self)
        for info in infos {
            dynamicValues[info.key.description] = Param(context: info)
        }
        _dynamics = Dynamics(wrappedValue: dynamicValues)
}
```

The `Param` is a simple property wrapper which expects a `String` as a type and a `ModelType`. I reworked the `func decode` of the `RequestInjectableDecoder` to allow the exporter access to the respective model type of the parameter. 
When the `Read` is called, it build an `[FieldKey: String] ` directory using the wrapped values of the `Param`. 

```
let queryInfo: [FieldKey: String] = dynamics
            .reduce(into: [FieldKey: String?](), { result, entry in
                result[Model.fieldKey(for: entry.key)] = entry.value.wrappedValue
            })
            .compactMapValues({ $0 })
```
It passes the directory to the `QueryBuilder` and executes the query using `queryBuilder.execute(on: database)` which returns the results of the query in a `EventLoopFuture`. Internally the `QueryBuilder` creates a `Fluent.Query` for every entry of the directory and accumulates them to a single query. This depends highly on the `Fluent` framework.

```
public func execute(on database: Fluent.Database) -> EventLoopFuture<[Model]> {
        
        func accumulateFilters(queryBuilder: inout Fluent.QueryBuilder<Model>, key: FieldKey, value: String, method: DatabaseQuery.Filter.Method) {
            if let intValue = Int(value) {
                queryBuilder.filter(key, method, intValue)
            } else if let boolValue = Bool(value) {
                queryBuilder.filter(key, method, boolValue)
            } else {
                queryBuilder.filter(key, method, value)
            }
        }
        
        var queryBuilder = Model.query(on: database)
        for (key, value) in parameters {
            accumulateFilters(queryBuilder: &queryBuilder, key: key, value: value, method: .equal)
        }
        return queryBuilder.all()
    }
```

### Points of Discussion / Problems
- Currently the Querybuilder has to cast the type to see if it can be an integer or a bool. Although we can find the type at runtime, I cannot see a way which would allow use to use a specific `Param` like `Param<Int>` instead of having all parameters conforming to type String
- Maybe the exporters need more information on how to decode. Currently the info they get is the FieldKey and that they should cast whatever they find to String.

The PR currently contains also some components for the other CRUD operations. They can either be eventually merged with the proposals or I could create separate PRs for these.

Commit:
First experiment with database components

Commit:
Merge branch 'develop' into feature/database-components

Commit:
moved query builder to separate file

Commit:
Merge branch 'develop' into feature/database-components

theMomax committed:
generalize mechanism used by RequestInjectable and extend it to DynamicProperty and Dynamics

theMomax committed:
remove unwanted change

Commit:
Merge branch 'develop' into feature/database-components

theMomax committed:
fix swiftlint

theMomax committed:
fix swiftlint 2

theMomax committed:
fix swiftlint 3

theMomax committed:
add assertion checking Dynamics is a struct

theMomax committed:
add tests

theMomax committed:
make Dynamics conform to ExpressibleByDictionaryLiteral

theMomax committed:
Merge branch 'develop' into feature/dynamic-property-and-dynamics

theMomax committed:
fix swiftlint again

theMomax committed:
fix ordering bug in execute test case

theMomax committed:
Merge branch 'develop' into feature/dynamic-property-and-dynamics

theMomax committed:
add @dynamicMemberLookup to Dynamics + typed<T>() function to retrieve typed array

theMomax committed:
make Dynamics generic (without restriction) + Dynamics conform to Collection

Commit:
prepare for presentation

theMomax committed:
Merge branch 'develop' into feature/dynamic-property-and-dynamics

Commit:
Merge branch 'feature/dynamic-property-and-dynamics' into feature/database-components

theMomax committed:
Merge branch 'develop' into feature/dynamic-property-and-dynamics

theMomax committed:
make Dynamics a property wrapper

Commit:
Merge branch 'feature/dynamic-property-and-dynamics' into feature/database-components

theMomax committed:
hide Runtime + fix type + fix SwiftLint

theMomax committed:
replace usage of Mirror by execute

theMomax committed:
fix usage of executable

Commit:
First draft of Get using Dynamics

Commit:
Merge branch 'feature/dynamic-property-and-dynamics' into feature/database-components

Commit:
First draft of a working query builder using Dynamics

Commit:
clean up

nerdsupremacist commented on Sources/Apodini/Components/Create.swift:
```diff
@@ -0,0 +1,143 @@
+import Fluent
+import Vapor
+
+public protocol DatabaseInjectionContext {
```
I love having clearly separate files if you don't mind. Please split up the files a bit

nerdsupremacist commented on Sources/Apodini/Components/Create.swift:
```diff
@@ -0,0 +1,143 @@
+    var type: Any.Type { get }
+}
+
+public struct Create<T: DatabaseModel>: Component where T.IDValue == UUID {
```
noob question: why does IDValue have to be a UUID?

nerdsupremacist commented on Sources/Apodini/Components/Create.swift:
```diff
@@ -0,0 +1,143 @@
+        for info in infos {
+            dynamicValues[info.key.description] = Param(context: info)
+        }
+        _dynamics = Dynamics(wrappedValue: dynamicValues)
```
Love it!

nerdsupremacist commented on Sources/Apodini/Components/Create.swift:
```diff
@@ -0,0 +1,143 @@
+    public func handle() -> EventLoopFuture<T> {
+        return T.find(id, on: database).flatMapThrowing({ model -> T in
+            model?.update(object)
+            return model!
```
please throw an error if the item does not exist

nerdsupremacist commented on Sources/Apodini/Components/Create.swift:
```diff
@@ -0,0 +1,143 @@
+    var _value: String?
+    
+    var wrappedValue: String? {
+        if context.key == FieldKey("id") || context.key == FieldKey("name") {
```
Why the special case here? My instinct tells me this is debugging code, but maybe I'm overseeing something

nerdsupremacist commented on Sources/Apodini/Request/Body.swift:
```diff
@@ -0,0 +1,35 @@
+
+
+@propertyWrapper
+public struct Body<Element: Codable>: RequestInjectable {
```
I thought we removed this. Maybe you added it again while rebasing?

nerdsupremacist reviewed (commented):
Great job! A lot of cool concepts here! and I love the interesting use of the new dynamics...
I'd like to merge this right after merging the dynamics PR.
Let's work on this during the Work Meeting ;)

I'd love to clean up the different property wrappers you implemented here, but I guess that most of them are necessary until all the param + vapor stuff is sorted out properly (is it fixed already I'm not sure)

Commit:
Removed debug code and code relicts

hendesi commented:
@nerdsupremacist Thanks for the feedback! Your instinct was right, most of it was debug code or some code relicts that somehow stayed there during the rebase. I removed them, moved different structs to different files and cleaned the code up a bit in general 👍

theMomax commented:
This is a pretty decent feature👍🏻. However, I think there is still quite some work to do in order to clean up the implementation:

In general I think that these database-related `Components` belong into a "Foundation" package and should thus not make use of **internal** elements (e.g. `RequestInjectable`).

As far as I can see you are not alone with the problem you encountered with `RequestInjectable` being not sufficient for your needs. I had the same problem with the old `RequestInjectable` (thanks to @tgymnich's revision it is now sufficient for me) and I think the REST-exporter is also struggling to get all the required information. I think we should have a bigger discussion on how to tackle this. An abstracted way of your requirements should also be covered by the result.

I think we should also adapt `@Parameter` to fit your needs to replace your new `Param`.

All of these changes to the core of Apodini should of course be rather abstracted and database-agnostic.

Commit:
Removed Param, create database property option to use Parameter

PSchmiedmayer reviewed (changes requested):
@hendesi I have looked over your PR but as far as I can see most of the elements that I would have commented on are dependent on the PR of @theMomax which should be merged very soon and is already addressed by some of the PRs that are linked to this PR such as #84 which looks like a good approach to me 👍

For now I would leave a detailed review of the PR but let me know as soon as the other PRs are merged and you have a comprehensive version of this PR that introduced the components and includes the functionality of the other PRs. 

In general I would suggest to add some documentation to the respective components to make clear what they can be used for and how a developer can use them in an Apodini project. In general I think that the ideas you have are going in the right direction and I am looking forward to what will be possible with the different database components. 
In addition I would suggest to move the different components such as `Read`, `Create`, ... to a separate Apodini target that we should call `ApodiniFoundation` similar to how the protocol buffer decoder is in a seperate target. This will make it more clear what public functionality will be needed to enable the usage of Apodini for other developers and will lead to a clear and well usable public interface of the Apodini project.

Commit:
Merge branch 'develop' into feature/database-components

Commit:
moved database components into seperate target

hendesi commented:
@PSchmiedmayer I moved the database files into a separate target. I'd like to get rid of the different fluent database drivers (e.g. `.product(name: "FluentMongoDriver", package: "fluent-mongo-driver")` in the Apodini target and move them to the Database target. This would impact the `DatabaseConfiguration` in which the Environment value of the database is set. 
```
if let database = app.databases.ids().map({ app.db($0) }).first {
                EnvironmentValues.shared.database = database
}
```
Since `shared` is currently internal, it's not possible to move the config out of main Apodini target. I personally don't think we should change the scope of shared. But this would leave us with having the driver dependencies in main Apodini.

Commit:
Renamed to ApodiniDatabase

Commit:
Merge branch 'develop' into feature/database-components

Commit:
Added descriptive comments

Commit:
Merge branch 'develop' into feature/database-components

PSchmiedmayer commented:
@hendesi I agree that the shared values should only be accessible internally but I would envision an API that would be public to manually add new values to the `EnvironmentValues.shared` instance. I think an addition only public API would be fine and in long term I could see how these shared environment values could move the the own `Application` type introduced in #82 which would also be available to any `Configuration` instances.

Commit:
Added database handler tests

Commit:
Fix swiftlint

Commit:
fix swiftlint

Commit:
Fix swiftlint

Commit:
fix swiftlint

Commit:
fix swiftlint

hendesi committed:
Merge branch 'develop' into feature/database-components

Commit:
Fix update handler test

Commit:
Merge branch 'feature/database-components' of https://github.com/Apodini/Apodini into feature/database-components

Commit:
Fix build

Commit:
fix test to pass on remote

Commit:
removed useless parameter mode

Commit:
Added testcase for querybuilder

hendesi committed:
Merge branch 'develop' into feature/database-components

hendesi commented:
After a lot of SwiftLint fixing and test adding, I think I implemented the requested changes. 
Changes I made: 
- Added description
- Moved Database into separate target
- Added tests for database handlers and querybuilder
- Removed unused parameter option

Re-requesting review again from @PSchmiedmayer @theMomax @nerdsupremacist 

@PSchmiedmayer I like the solution of offering a public API for developers to add environment values by themselves. But I would move this to a separate PR and try to get this one merged first if you agree 👍

hendesi committed:
Merge branch 'develop' into feature/database-components

PSchmiedmayer commented on Sources/ApodiniDatabase/DatabaseModel.swift:
```diff
@@ -0,0 +1,15 @@
+import Vapor
+
+///A protocol all Models that are used with `ApodiniDatabase` need conform to
+public protocol DatabaseModel: Content, Model where IDValue == UUID {
```
What is the reason that the `IDValue` is constrained to `UUID`. What would happen if the ID value is something else?

hendesi replied:
This was something I've thought about before we introduced the new Parameter retrieval. The reason was that for `Model.IDValue` needs to conform to `LosslessStringConvertible`, which it does now anyway. So I'll remove that 👍

PSchmiedmayer commented on Sources/ApodiniDatabase/Handlers/Create.swift:
```diff
@@ -0,0 +1,30 @@
+            }
+            .wait()
+        
+        return try! T
```
What is the reasoning behind fetching the type from the database? Would it be fine to store `object` to a local property and return it after saving the element to the database. I think as `object` is an instance of a class type the ID should also be updated after the save call.

hendesi replied:
I think im not quite sure what you mean. This code is like it is due to `EventLoopFutures` not being a possible return type. Once this is possible again `Create` would return an `EventLoopFuture<Model>, where `Model` is the saved object. If we would store `object` to a local property and return it after the database save, we would again have some kind of blocking. But I agree with you that the returned object should have the saved id, I'll add that

PSchmiedmayer replied:
Sorry for the confusing, I could have been clearer with my comment. Here is an example of what I have meant with leaving out the fetching the type from the database.

Fluent automatically updates the ID property on the save call. So you could have a code similar to this one that does both the saving and the returning the object in a single call as seen in the following code block:
```swift
object
    .save(on: database)
    .transform(to: object)
    .wait()
```

Once we have merged #94 or #98 or a mixture of both you can simply remove the `wait()` call and the method should return the instance include the ID that has been created from the database.

[The Vapor template does something similar but uses `map` instead of `transform`](https://github.com/vapor/template/blob/0f3b552f66aee092b457a542d9a83dd5578d05b6/Sources/App/Controllers/TodoController.swift#L20) which has the same effect without the need for a closure in this context.

hendesi replied:
Thanks for clarifying! I agree, that looks a lot cleaner. I will update the `Create` Handler accordingly.

PSchmiedmayer commented on Sources/ApodiniDatabase/Handlers/Create.swift:
```diff
@@ -0,0 +1,30 @@
+            .map { _ in
+                self.object
+            }
+            .wait()
```
I would suggest to wait for the merging of #94 or #98 or a mix between the two ideas to we don't introduce blocking code in develop. It would also be great to get your input in the ideas as well as the #97 as your code would be one of the main users of asynchronous code.

PSchmiedmayer commented on Sources/ApodiniDatabase/Handlers/Create.swift:
```diff
@@ -0,0 +1,30 @@
+    private var object: T
+
+    public func handle() -> T? {
+        // swiftlint:disable all
```
Please don't disable all SwiftLint warnings. Please only disable a single rule and provide a detailed description on why the rule is disabled in a comment [as described in this message](https://github.com/Apodini/Apodini/pull/83#issuecomment-752128774)

In general I think it would be good to support throwing handlers that would be able to forward throwing errors to the client. Vapor supports this mapping of errors to networking errors and correct responses to the client. I think this would be a great addition to Apodini and the database code would be a prime example of the usage of throwing errors.

PSchmiedmayer commented on Sources/ApodiniDatabase/Handlers/Delete.swift:
```diff
@@ -0,0 +1,28 @@
+    @Parameter(.http(.path))
+    var id: UUID
+    
+    var idParameter: Parameter<UUID> {
```
As far as I can see the `idParameter` is used for testing purposes only. You could achieve the same result using reflection (`Mirror`) in the test case which would reduce the internal interface of the type.

PSchmiedmayer commented on Sources/ApodiniDatabase/Handlers/Read.swift:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    public func handle() -> String {
+//    public func handle() -> [Model] {
```
I think the current implementation doesn't return the actual `Models` as this functionality is commented out.

hendesi replied:
Some of the handlers currently return `String` because calling `wait()` on the `EventloopFuture` returned by the database operation results in an failed precondition as one is not allowed to call `wait()` on an `eventloop`.

```
Precondition failed: BUG DETECTED: wait() must not be called when on an EventLoop.
```

As you already suggested, it would beneficial to wait the merger of #94 or #98. If `EventloopFuture` works again, I would of course change it. So if you agree, I'd suggest we leave it like that for now.

PSchmiedmayer replied:
#98 should be merged pretty soon so I would be optimistic to have `EventLoopFuture` as possible `Handler` return types by tomorrow's team meeting 👍

PSchmiedmayer commented on Sources/ApodiniDatabase/Handlers/Read.swift:
```diff
@@ -0,0 +1,39 @@
+///     Read<Bird>()
+/// }
+/// Sending a request to  ~/api/birds/birds?age=19&name=Foo would return an array of `Bird` object that have an age of 19 and the name Foo.
+public struct Read<Model: DatabaseModel>: Handler {
```
I like the flexibility of the `Read` component using `@Properties` 🚀

PSchmiedmayer commented on Sources/ApodiniDatabase/Handlers/Update.swift:
```diff
@@ -0,0 +1,37 @@
+    @Parameter(.http(.path))
+    var id: UUID
+    
+    var idParameter: Parameter<UUID> {
```
See my comment in the `Delete` `Handler`

PSchmiedmayer commented on Sources/ApodiniDatabase/Handlers/Update.swift:
```diff
@@ -0,0 +1,37 @@
+    
+//    public func handle() -> EventLoopFuture<T> {
+    public func handle() -> String {
+        // swiftlint:disable all
```
Please rework the usage of `// swiftlint:disable` across the PR

PSchmiedmayer commented on Sources/ApodiniDatabase/Handlers/Update.swift:
```diff
@@ -0,0 +1,37 @@
+    @Apodini.Environment(\.database)
+    private var database: Fluent.Database
+    
+    @Parameter
```
I think updating would be an other great use cases of `@Properties` and a dynamic creation of these properties. E.g. we could achieve an `Update` component that can update a `Model` with a request that just provides one property of the `Model` and leaves all other elements like they are. I think this would be a very sophisticated usage of reflection and the runtime information and `@Properties` but would showcase the great flexibility of our approaches. I would be fine with this not being part of this PR but a subsequent PR. If this is the case pleas document the feature using a GitHub issue.

PSchmiedmayer commented on Sources/ApodiniDatabase/Query/QueryBuilder.swift:
```diff
@@ -0,0 +1,124 @@
+    
+    private func extract(from queryString: String) -> [FieldKey: String] {
+        var foundParameters: [FieldKey: String] = [:]
+        let extractedQueryString = queryString.split(separator: "?")[1]
```
I would suggest to check the count or possible indexes before accessing a collection using a fixed index.

PSchmiedmayer commented on Sources/ApodiniDatabase/Query/QueryBuilder.swift:
```diff
@@ -0,0 +1,124 @@
+    }
+    
+    // swiftlint:disable:next todo
+    //TODO: Find a better way to do this
```
I would suggest to implement some small visitor pattern with an extension to `FieldProperty` that would allow you to access the type and the corresponding information using type checks instead of string representations, let me know if you need some more input on this 👍

hendesi replied:
Its definitely a good idea to move away from checking string representations! 👍 Would your approach also allow to have some type info already on compile time? To have a real benefit, we would need to have the type info at compile time such that `Read` can initialise its `@Properties` with `Parameter<Type` and stop using `Parameter<String>` for everything

Maybe this is also something @nerdsupremacist could provide some input on 👍

theMomax replied:
I agree it would be really beneficial to have proper type-info here! While it does not make that much of a difference for HTTP url parameters, it does for other exporters which rely on JSON or protobuf. Also, it would look a bit ugly in the OpenAPI docs.

PSchmiedmayer replied:
@hendesi Yes, I could image that a visitor pattern based approach would be able to address this issue and would be able to get the type info on compile time. @nerdsupremacist or I can take a look at this in the work meeting and provide some input on how to implement this 🚀

hendesi replied:
@PSchmiedmayer Thanks! I already implemented the visitor pattern on my local branch and it works quite well. Might have some additional questions which we could discuss in the work meeting 👍

PSchmiedmayer reviewed (changes requested):
@hendesi Thanks for the updated PR for the Database components. I really like how you were able to split out the implementation into a separate target.

Regarding [your comment](https://github.com/Apodini/Apodini/pull/70#issuecomment-755298583): 
> I like the solution of offering a public API for developers to add environment values by themselves. But I would move this to a separate PR and try to get this one merged first if you agree 👍
I think it is fine to create a separate PR for this. Please create a GitHub issue for this task so we will be able to keep it in mind and track its progress.

I had a detailed look at the code and added some comments throughout the files. I think the biggest comment impacting this PR is my suggestion to wait for the merger of #94 or #98 or a mix between the two ideas. It be great to get your input and review of the PRs and what your comments regarding the different approaches are.
Apart from this I had some suggestions for some improvements and changes to the usage of the `swiftlint:disable` comments. Feel free to let me know if you have any questions or I can help you with anything. 

I am looking forward to seeing these features and `Handlers` in action! 🚀

theMomax commented on Sources/ApodiniDatabase/DatabaseModel.swift:
```diff
@@ -0,0 +1,15 @@
+internal extension DatabaseModel {
+    static func fieldKey(for string: String) -> FieldKey {
+        // swiftlint:disable:next force_unwrapping
+        Self.keys.first(where: { $0.description == string })!
```
Please provide an explanation if you do a force-unwrap. E.g.:
```suggestion
        if let key = Self.keys.first(where: { $0.description == string })! {
            return key
        } else {
            fatalError("Unexpectedly found nil, this shouldn't have happened as <reason>")
            exit(1)
        }
```

theMomax commented on Sources/ApodiniDatabase/Handlers/Read.swift:
```diff
@@ -0,0 +1,39 @@
+import Fluent
+import Apodini
+
+/// A Handler that returns an array of `DatabaseModel` which all conform to the query parameters of the request. It accepts all parameters that represent properties of the specified `DatabaseModel`.
```
Sorry for being that picky here, but since there is an open change-request anyways, please add line-breaks after ~150 characters on block-comments.

theMomax reviewed (commented):
In general, this looks very promising. It is really unfortunate how you suffer from `EventLoopFuture`s being dysfunctional currently. Except for smaller code-style issues, the only thing remaining is the missing type-info. I hope @nerdsupremacist can help you with that and I would also be happy to support you during the next work-meeting.

Commit:
Implemented feedback

Commit:
Merge branch 'feature/database-components' of https://github.com/Apodini/Apodini into feature/database-components

Commit:
Fix swiftlint

Commit:
Fix swiftlint

hendesi commented:
>  I hope @nerdsupremacist can help you with that and I would also be happy to support you during the next work-meeting.

Thanks! I think this could be really beneficial 👍

hendesi committed:
Merge branch 'develop' into feature/database-components

Commit:
First draft of type inference

PSchmiedmayer commented:
@hendesi We have merged #98 which now enables `EventLoopFutures` on the develop branch 🚀

Commit:
Added support for partial update handler

Commit:
Introduced visitor pattern to infer types

Commit:
Introduced visitor pattern to infer types

Commit:
Merge branch 'develop' into feature/database-type-inference

Commit:
Converted database handlers back to using eventloopfutures

Commit:
Cleanup

Commit:
removed unused class and added tests

Commit:
Reworked database folder structure

Commit:
Fix swiftlint

Commit:
fix swiftlint + added tests for updater

hendesi committed:
Merge branch 'develop' into feature/database-components

Commit:
fix swiftlint

Commit:
Merge branch 'feature/database-components' of https://github.com/Apodini/Apodini into feature/database-components

hendesi commented:
@PSchmiedmayer @nerdsupremacist @theMomax 
I added the following improvements to this PR:

- Added a visitor pattern to infer the types in the `Querybuilder` instead of using a string. 
- Modified the update handler to enable single properties updates
- Readded the eventloop returns of the handlers

Could you please take another look at this? So we can get this merged 👍

Commit:
fix swiftlint

Commit:
swiftlint fix

Commit:
fix build

PSchmiedmayer committed:
Merge branch 'develop' into feature/database-components

theMomax commented on Sources/Apodini/Configurations/DatabaseConfiguration.swift:
```diff
@@ -37,6 +37,7 @@ public final class DatabaseConfiguration: Configuration {
             let factory = try databaseFactory(for: self.type)
             databases.use(factory, as: databaseID)
             app.migrations.add(migrations)
+            print(migrations[0].name)
```
Is this intended as a startup log or is it a lost debug log?

hendesi replied:
lost debug log, should be removed 👍

theMomax commented on Sources/Apodini/Semantic Model Builder/REST/RESTInterfaceExporter.swift:
```diff
@@ -165,7 +164,16 @@ class RESTInterfaceExporter: InterfaceExporter {
+                     This is currently unsupported.
+                """)
+            do {
+                return try request.content.decode(Type.self, using: JSONDecoder())
```
This change seems a bit weird to me as it has nothing to do with Database components. Did this slip in from a merge?

hendesi replied:
This is due to having two Parameter properties in the `Update` handler of which one is nil. Currently there is no support for requests contain more than one body parameter. If the request contains something like
```
let parameters: [String: TypeContainer] = [
            "name": TypeContainer(with: "FooBird")
]
```
It would try and decode it as a `Bird` which would fail and throw a `DecodingError`. But as this can also be decoded as `Parameter<[String: TypeContainer]>, it should fail silently

theMomax replied:
Ah, makes sense, thanks for the explanation!

theMomax commented on Sources/Apodini/WebService.swift:
```diff
@@ -4,10 +4,9 @@
 //  Created by Paul Schmiedmayer on 7/6/20.
 //
-
+import Fluent
```
This can be removed since there is no other change in that file, right?

theMomax commented on Sources/ApodiniDatabase/Handlers/Update/Updater.swift:
```diff
@@ -0,0 +1,58 @@
+                                _ = model.update(on: database).transform(to: model)
+                                return model
+                            } else {
+                                fatalError("the updater was unable to update single properties of")
```
Error message seems to be incomplete.

hendesi committed:
Merge branch 'develop' into feature/database-components

theMomax commented on Sources/ApodiniDatabase/Handlers/Update/Updater.swift:
```diff
@@ -0,0 +1,58 @@
+                .map { foundModel -> Model in
+                    foundModel.update(model)
+                    _ = foundModel.update(on: database).transform(to: model)
+                    return model
```
That means as soon as one provides the `model`, all `properties` are ignored, right? Just from the endpoint's API I would have guessed `properties` to not be ignored, but to overwrite their counterparts on `model` if `model` is provided. I isn't that important, but I expected it to behave differently. Maybe just emphasize this a bit more on the `Update`'s docs.

hendesi replied:
> That means as soon as one provides the model, all properties are ignored, right?

Yes that is correct. I'll update the docs of the `Update` handler to clarify things a bit there.

theMomax commented on Tests/ApodiniTests/DatabaseTests/QueryBuilderTests.swift:
```diff
@@ -0,0 +1,18 @@
+        
+        let birds = try queryBuilder.execute(on: app.db).wait()
+        XCTAssert(birds.count == 1)
+        XCTAssert(birds[0].name == "Swift")
```
```suggestion
        XCTAssert(birds[0].name == "Swift")
        XCTAssert(birds[0].age == 5)
```

theMomax commented on Sources/ApodiniDatabase/Context/TypeContainer.swift:
```diff
@@ -0,0 +1,186 @@
+
+extension TypeContainer: LosslessStringConvertible {
+    public init?(_ description: String) {
+        self = .noValue
```
What is this used for? What implications does it have that you always return `.noValue`?

hendesi replied:
Since all query parameters need to conform to `LosslessStringConvertible`, but are then internally rated as `.lightweight` and there for not initialised using `init(_ description: String)`, there is currently no need to implement some decision pattern here to init the right type.
But I see your point, this should be handled eventually.

theMomax replied:
Ok, just add a comment with this rationale to the code please.

PSchmiedmayer replied:
I would suggest adding a comment and using a `fatalError(...)` with a message that would clearly indicate why this initializer could never be called.

theMomax reviewed (commented):
Mostly looks really good! I noted a few smaller things. I dislike the huge switch statements, but I guess they are necessary. Only bigger thing: Optionals aren't supported, are they?

Commit:
Implemented feedback

hendesi commented:
> I dislike the huge switch statements, but I guess they are necessary. 

@theMomax  I agree. But its something which is somewhat necessary in the code, at least as far as I know. If we can come up with something more elegant, we can definitely update this section. But for now I think, it is good enough for a first version.

I addressed all of your comments. Could you please have a final look? 👍

theMomax commented on Sources/ApodiniDatabase/Handlers/Update/Updater.swift:
```diff
@@ -41,7 +41,7 @@ internal struct Updater<Model: DatabaseModel> {
                                 return model
                             } else {
-                                fatalError("the updater was unable to update single properties of")
+                                fatalError("An error occurred while updating single parameters the model ")
```
"of" missing 🙃

theMomax reviewed (approved):
Looks good from my point of view!

Commit:
Added comment on init + typo

Commit:
added testcase for typecontainer

Commit:
Refined testcase for type container

Commit:
Merge branch 'develop' into feature/database-components

hendesi committed:
Merge branch 'develop' into feature/database-components

hendesi commented:
@PSchmiedmayer  @nerdsupremacist Could you please have another look at this? 
I would really appreciate it if we can get this merged soon  👍

PSchmiedmayer commented on Sources/Apodini/Semantic Model Builder/REST/RESTInterfaceExporter.swift:
```diff
@@ -165,7 +164,16 @@ class RESTInterfaceExporter: InterfaceExporter {
+                return try request.content.decode(Type.self, using: JSONDecoder())
+            } catch {
+                return nil
+            }
```
I would suggest to add a `try?` operator here:
```suggestion
        return try? request.content.decode(Type.self, using: JSONDecoder())
```

PSchmiedmayer commented on Sources/Apodini/Semantic Model Builder/REST/RESTInterfaceExporter.swift:
```diff
@@ -165,7 +164,16 @@ class RESTInterfaceExporter: InterfaceExporter {
+                     A Handler could define multiple .content Parameters. In such a case the REST exporter would
+                     need to decode the content via a struct containing those .content parameters as properties.
+                     This is currently unsupported.
+                """)
```
Please remove warning found in the source code. I would suggest to add a detailed issue about the problem and maybe add a comment here that describes the issue and references the GitHub issue.

PSchmiedmayer commented on Sources/ApodiniDatabase/Context/DatabaseInjectionContext.swift:
```diff
@@ -0,0 +1,22 @@
+    
+    static func == (lhs: ModelInfo, rhs: ModelInfo) -> Bool {
+        lhs.key == rhs.key
+    }
```
As far as I can see the conformance to `Equatable` is not need at the moment.

PSchmiedmayer commented on Sources/ApodiniDatabase/DatabaseModel.swift:
```diff
@@ -0,0 +1,23 @@
+import Fluent
+import Vapor
```
Please make sure that Vapor is only implemented as `@_implementationOnly`.
As far as I can see an Foundation import would be sufficient:
```suggestion
import Foundation
```
If you remove the `description` property you can even remove the import.

PSchmiedmayer commented on Sources/ApodiniDatabase/DatabaseModel.swift:
```diff
@@ -0,0 +1,23 @@
+        if let key = Self.keys.first(where: { $0.description == string }) {
+            return key
+        } else {
+            fatalError("Unexpectedly found nil. Failed to find a Fieldkey under the given string.")
```
I would suggest to add the name of the key using string interpolation to the error description to make it easier to identify the issue when encountering the error.

PSchmiedmayer commented on Sources/ApodiniDatabase/DatabaseModel.swift:
```diff
@@ -0,0 +1,23 @@
+    
+    static var description: String {
+        NSStringFromClass(Self.self)
+    }
```
As far as I can see the property is not needed. I would suggest removing it.

PSchmiedmayer commented on Sources/ApodiniDatabase/Handlers/Delete.swift:
```diff
@@ -0,0 +1,20 @@
+        Model.find(id, on: database)
+            .unwrap(orError: Abort(.notFound) )
+            .flatMap { $0.delete(on: database ) }
+            .transform(to: HTTPStatus.ok.code )
```
I think just returning a `Void` would be sufficient. We will need some way to indicate a HTTP status code for a response. I would suggest to add an GitHub issue to document the need for such a feature and maybe think about a way to return a response as well as an appropriate status code.
I would suggest to just remove nothing in the case the deletion was successful.

hendesi replied:
Not sure if I'm missing something, but it's currently not possible to return `EventloopFuture<Void>` as a `Response`, because Void can't conform to `ResponseTransformable`

PSchmiedmayer replied:
Ah, I see. We should probably add an implementation to transform `Void` to a response `.final(Void())`. I would suggest to create an issue about this to discuss if the decisions to return a `.final(Void())` would be correct. An alternative would be to return `.nothing` but this would not create any response to the client which would be needed here, therefore I would not suggest to use `.nothing` here.

PSchmiedmayer replied:
Ah, I see. We should probably add an implementation to transform `Void` to a response `.final(Void())`. I would suggest to create an issue about this to discuss if the decisions to return a `.final(Void())` would be correct. An alternative would be to return `.nothing` but this would not create any response to the client which would be needed here, therefore I would not suggest to use `.nothing` here.

PSchmiedmayer commented on Sources/TestWebService/main.swift:
```diff
@@ -122,6 +121,7 @@ struct TestWebService: Apodini.WebService {
     }
 
     @PathParameter var userId: Int
+    @PathParameter var dummy: String
```
As far as I can see this `@PathParameter` can be removed.

PSchmiedmayer commented on Sources/TestWebService/main.swift:
```diff
@@ -158,6 +158,10 @@ struct TestWebService: Apodini.WebService {
+    
+    var configuration: Configuration {
+        HTTP2Configuration()
+    }
```
I am not sure if this added configuration is relevant for the PR

PSchmiedmayer commented on Sources/TestWebService/main.swift:
```diff
@@ -4,9 +4,8 @@
-
 import Apodini
-
+import ApodiniDatabase
```
Feel free to add some examples `ApodiniDatabase` examples in the `TestWebService` 👍

PSchmiedmayer replied:
I there is no example in the `TestWebService` I would suggest to remove the import.

PSchmiedmayer commented on Sources/ApodiniDatabase/Handlers/Read/QueryBuilder.swift:
```diff
@@ -0,0 +1,104 @@
+internal struct QueryBuilder<Model: DatabaseModel> {
+    internal var debugDescription: String {
+        parameters.debugDescription
+    }
```
Doesn't seem to be used, I would consider removing it.

PSchmiedmayer commented on Sources/ApodiniDatabase/Handlers/Read/QueryBuilder.swift:
```diff
@@ -0,0 +1,104 @@
+
+    private var fieldKeys: [FieldKey] {
+        type.keys
+    }
```
Doesn't seem to be used, I would consider removing it.

PSchmiedmayer commented on Sources/ApodiniDatabase/Handlers/Read/QueryBuilder.swift:
```diff
@@ -0,0 +1,104 @@
+    @discardableResult
+    internal static func fieldKeys<Model: DatabaseModel>(for type: Model.Type) -> [FieldKey] {
+        type.keys
+    }
```
Doesn't seem to be used, I would consider removing it.

PSchmiedmayer commented on Sources/ApodiniDatabase/Handlers/Read/Read.swift:
```diff
@@ -0,0 +1,40 @@
+/// }
+/// Sending a request to  ~/api/birds/birds?age=19&name=Foo would 
+///return an array of `Bird` object that have an age of 19 and the name Foo.
+public struct Read<Model: DatabaseModel>: Handler {
```
Looks nice to me! I like the longer documentation on top. It would be interesting to also have a `ReadOne` or `ReadSingle` or similar `Handler` that only returns a single instance based on an id passed in using a path parameter.

PSchmiedmayer commented on Sources/ApodiniDatabase/Handlers/Update/Updater.swift:
```diff
@@ -0,0 +1,58 @@
+    func executeUpdate(on database: Database) -> EventLoopFuture<Model> {
+        if let model = model {
+            return Model.find(modelID, on: database)
+                .unwrap(orError: Abort(.notFound))
```
This seems to be a duplicated from the else case. I would suggest to split up the method into two smaller ones that split up after this point to make the method more readable and then you can also document the decision mentioned in the [comment by @theMomax below](https://github.com/Apodini/Apodini/pull/70/files#r559168012) in each of the method documentations.

PSchmiedmayer commented on Sources/ApodiniDatabase/Handlers/Update/Updater.swift:
```diff
@@ -0,0 +1,58 @@
+                .map { foundModel -> Model in
+                    foundModel.update(model)
+                    _ = foundModel.update(on: database).transform(to: model)
+                    return model
```
This call returns before the model is actually updated. I would suggest to use a `flatMap` here and return the future from the `.transform(to: model)` call:
```suggestion
                .flatMap { foundModel -> EventLoopFuture<Model> in
                    foundModel.update(model)
                    return foundModel
                        .update(on: database)
                        .transform(to: model)
                }
```

PSchmiedmayer commented on Sources/ApodiniDatabase/Handlers/Update/Updater.swift:
```diff
@@ -0,0 +1,58 @@
+                        let fieldKey = Model.fieldKey(for: label.trimmed())
+                        if let value = properties[fieldKey] {
+                            if visitable.accept( ConcreteUpdatableFieldPropertyVisitor(updater: value) ) {
+                                _ = model.update(on: database).transform(to: model)
```
This method also returns before the model is actually saved to the database. Similar to above you will need to use `flatMap` here. Otherwise errors when saving an element to the database would get lost in the void.

PSchmiedmayer commented on Sources/ApodiniDatabase/Handlers/Update/Updater.swift:
```diff
@@ -0,0 +1,58 @@
+                    for child in Mirror(reflecting: model).children {
+                        guard let visitable = child.value as? UpdatableFieldProperty,
+                              let label = child.label
+                        else { continue }
```
This method is overall very hard to read. Maybe this functionality can be moved to an extension of the `Model` class? In addition it is hard to see the guard statement here. I would suggest a different formatting:
```suggestion
                        guard let visitable = child.value as? UpdatableFieldProperty,
                              let label = child.label else { 
                            continue
                        }
                        
```

PSchmiedmayer commented on Sources/ApodiniDatabase/Handlers/Update/Updater.swift:
```diff
@@ -0,0 +1,58 @@
+}
+
+fileprivate extension String {
+    func trimmed() -> Self {
```
I would suggest to add a more descriptive name and some documentation and examples on why this is necessary.

PSchmiedmayer commented on Sources/ApodiniDatabase/Handlers/Update/Updater.swift:
```diff
@@ -0,0 +1,58 @@
+                                _ = model.update(on: database).transform(to: model)
+                                return model
+                            } else {
+                                fatalError("An error occurred while updating single parameters of the model ")
```
Please add a more descriptive error message that e.g. includes the `fieldKey` and more context information.

PSchmiedmayer commented on Sources/ApodiniDatabase/PropertyVisitors/UpdatableFieldProperty.swift:
```diff
@@ -0,0 +1,37 @@
+
+/// A protocol all field property visitors that want to update values of a `FieldProperty` have to conform to.
+/// It returns whatever has been specified as `Value`.
+protocol UpdatableFieldPropertyVisitor where Value: Codable {
+    associatedtype Value
+
+    func visit<Model, V>(_ property: inout FieldProperty<Model, V>) -> Value
+}
+
+/// A protocol to make the `FieldProperty` of a database model updatable at run time.
+/// It returns `true` if the update of the given `FieldProperty` was successful, else `false`
+protocol UpdatableFieldProperty {
+    func accept<Visitor: UpdatableFieldPropertyVisitor>(_ visitor: Visitor) -> Visitor.Value where Visitor.Value == Bool
+}
```
As far as I can see there is no other implementation then the one defined in this scope and all use `Bool` as the concrete type for `Value`. Therefore I would suggest to remove the `associatedtype` completely and only use concrete types.

PSchmiedmayer commented on Sources/ApodiniDatabase/PropertyVisitors/UpdatableFieldProperty.swift:
```diff
@@ -0,0 +1,37 @@
+
+/// A concrete implementation of the `UpdatableFieldPropertyVisitor` that updates a given `FieldProperty` and returns an `true` if the update was successful
+struct ConcreteUpdatableFieldPropertyVisitor: UpdatableFieldPropertyVisitor {
+    typealias Value = Bool
```
Can be removed if the comment above is applied.

PSchmiedmayer commented on Sources/ApodiniDatabase/PropertyVisitors/VisitableFieldProperty.swift:
```diff
@@ -0,0 +1,57 @@
+
+/// A protocol all field property visitors have to conform to. It returns whatever has been specified as `Value`.
+protocol FieldPropertyVisitor where Value: Codable {
+    associatedtype Value
```
The `associatedtype` is not needed in the current implementation as far as I can see as you only use the concrete type `TypedContainer` everywhere. I would suggest removing it to reduce the complexity of reading the code and steering the type checker a bit less 😅

PSchmiedmayer commented on Sources/ApodiniDatabase/PropertyVisitors/VisitableFieldProperty.swift:
```diff
@@ -0,0 +1,57 @@
+struct ConcreteTypeVisitor: FieldPropertyVisitor {
+    typealias Value = TypeContainer
+    
+    func visit<Model, V>(_ property: FieldProperty<Model, V>) -> Value where Model: Fields, V: Decodable, V: Encodable {
```
You can combine the `V: Decodable, V: Encodable` requirement into a single one. In addition you don't need the `typealias` here if you remove the `associatedType` above.
```suggestion
    func visit<Model, V>(_ property: FieldProperty<Model, V>) -> TypeContainer where Model: Fields, V: Codable {
```

PSchmiedmayer commented on Sources/ApodiniDatabase/PropertyVisitors/VisitableFieldProperty.swift:
```diff
@@ -0,0 +1,57 @@
+struct ConcreteIDPropertyVisitor: IDPropertyVisitor {
+    typealias Value = TypeContainer
+    
+    func visit<Model, V>(_ property: IDProperty<Model, V>) -> TypeContainer where Model: DatabaseModel, V: Decodable, V: Encodable {
```
Same comment as found in `ConcreteTypeVisitor`.

PSchmiedmayer reviewed (changes requested):
@hendesi I have taken a closer look at the PR and had a few comments here and there about simplifications and some `EventLoopFuture` improvements in the code. 
Sometimes I have added a comment about unused code, feel free to let me know if I missed some usages of the code or if this will be come essential for some additional features that build on top of the existing implementation.

In addition I would suggest to take a look at the [automated comments from SonarCloud]( https://sonarcloud.io/project/issues?id=Apodini_Apodini&pullRequest=70&resolved=false&types=CODE_SMELL) that might also be interesting and helpful for improving the PR.

Thanks for already decoupling the implementation into a separate target 🚀
The overall functionality and feature set seems great to me, really looking forward to seeing this feature in Apodini. Thanks for all the work that went into the PR and the long wait until `EventLoopFutures` were added back into develop.

Commit:
Implemented feedback by paul

hendesi committed:
Merge branch 'develop' into feature/database-components

hendesi commented:
@PSchmiedmayer Thanks for the detailed feedback! I think I addressed all of it. 👍 
I have only one [comment](https://github.com/Apodini/Apodini/pull/70#discussion_r560055814) regarding the `Delete` handler.
Other than that, I definitely want to add a `ReadSingle` Handler, that is a good idea. But I would vote for moving this to a separate pr. The focus should be to have a functional version of the database in develop and not to put every feature into one PR 😅 .

PSchmiedmayer commented on Sources/ApodiniDatabase/Context/TypeContainer.swift:
```diff
@@ -184,5 +184,6 @@ extension TypeContainer: LosslessStringConvertible {
         // As query parameters are currently internally used a `.lightweight` and therefore not initialized using this init,
         // there is currently no mapping for this and will always be defaulted to `.noValue`.Will be added in the future.
         self = .noValue
```
You can remove this assignment to `self` as the initializer will not terminate due to the `fatalError` below.

PSchmiedmayer commented on Sources/ApodiniDatabase/Context/DatabaseInjectionContext.swift:
```diff
@@ -0,0 +1,18 @@
+}
+
+///A struct implementing `DatabaseInjectionContext` and containing a fieldkey and the expected type for that key.
+struct ModelInfo: DatabaseInjectionContext, Equatable {
```
I would suggest to remove the conformance to `Equatable` as there is no implementation in the type anymore and it doesn't seem to be sued anywhere.

PSchmiedmayer reviewed (approved):
@hendesi Thanks for incorporating all the feedback! I agree that the `SingleRead` should be a separate PR. I would suggest to create multiple issues from this PR:

1. The issue with multiple currently conflicting parameters for the `RESTInterfaceExporter` that you previously commented using a warning
2. The issue with returning `Void` from a `handle()` function and how to return different status codes then `200` as raised by the [thread and issue in the `Delete` component](https://github.com/Apodini/Apodini/pull/70/files/00759a1d3bf1f57b10ed377218010dd67d9848e0#r559817626).

Apart from this I only had a few smaller comments. I would already approve the PR as I don't think there is a need for an additional review after the issues are created and the small comments are addressed 👍

Commit:
Final clean up

Commit:
Merge branch 'feature/database-components' of https://github.com/Apodini/Apodini into feature/database-components

PSchmiedmayer commented:
In addition please take a look at the issues detailed by SonarCloud that might be resolved before merging the PR: https://sonarcloud.io/project/issues?id=Apodini_Apodini&pullRequest=70&resolved=false&types=CODE_SMELL

Commit:
fix sonarcloud warnings
